### PR TITLE
Fix replay divergence with LeakSanitizer

### PR DESCRIPTION
**Pull request by Keno**
LeakSanitizer reads all writable memory regions of loaded shared
libraries to find any pointer references that are still live
at the end of the process (in order to be able to complain about
the ones that are not, but were also not freed). Unfortunately,
at the moment, this causes it to read the globals.in_replay flag,
causing a divergence. Fix that by moving the in_replay flag to
the rr page, which already differs between record/replay anyway
(and is already excluded from scanning by LeakSanitizer,
because it's not marked PF_W).

### CODE REVIEW

**Comment by Keno** [comment]
Actually, the same issue applies to breakpoint_value, so I'll have to move that also.

**Comment by Keno** [comment]
> Actually, the same issue applies to breakpoint_value, so I'll have to move that also.

Just to keep track, I did that and the lsan test case works now. I didn't update the test case included here, because it's not super easy to predict when breakpoint_value would have generally been set, and it didn't always trigger with lsan either, but I think the test case will help prevent future regressions of people introducing divergent memory content into the PF_W mappings of any `.so`s.

File: src/util.cc [hunk 1]
```diff
@@ -446,10 +446,11 @@ static void iterate_checksums(Task* t, ChecksumMode mode,
...
   }
 
+  ReplaySession *replay = t->session().as_replay();
   remote_ptr<unsigned char> in_replay_flag;
   unsigned char in_replay = 0;
-  if (t->preload_globals) {
-    in_replay_flag = REMOTE_PTR_FIELD(t->preload_globals, in_replay);
+  if (replay && replay->has_trace_quirk(TraceReader::UsesGlobalsInReplay) && t->preload_globals) {
+    in_replay_flag = REMOTE_PTR_FIELD(t->preload_globals, reserved_legacy_in_replay);
     in_replay = t->read_mem(in_replay_flag);
```
Comment (rocallahan) [inline src/util.cc:454]: How can `in_replay` be anything other then 1 here?

Comment (Keno) [reply src/util.cc:454]: I suppose it can't anymore. I'll remove the memory read.

**Comment by Keno** [comment]
I've updated this to address the review comment. Sorry about the delay.

**Pull request merged**